File: common/src/crypto.rs
use chacha20poly1305::{
    aead::{Aead, KeyInit, OsRng},
    ChaCha20Poly1305, Nonce,
};
use hkdf::Hkdf;
use rand::RngCore;
use sha2::{Sha256, Digest};
use zeroize::{Zeroize, ZeroizeOnDrop};
use base64::{Engine as _, engine::general_purpose::URL_SAFE_NO_PAD};

/// Generatore di codici chat sicuri (256-bit, base64url)
pub fn generate_chat_code() -> String {
    let mut bytes = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut bytes);
    URL_SAFE_NO_PAD.encode(bytes)
}

/// Generatore di codici chat semplici (6 cifre numeriche)
/// ATTENZIONE: Meno sicuro del formato completo (solo ~20 bit di entropia vs 256 bit)
pub fn generate_numeric_chat_code() -> String {
    use rand::Rng;
    let code = rand::thread_rng().gen_range(100000..=999999);
    format!("{:06}", code)
}

/// Genera un identificatore di chat che il server può usare senza conoscere il codice originale
/// Il server usa questo hash per identificare la chat, ma non può derivare la chiave E2EE
pub fn chat_code_to_room_id(chat_code: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(b"rchat-room-id-v1:");
    hasher.update(chat_code.as_bytes());
    let hash = hasher.finalize();
    URL_SAFE_NO_PAD.encode(hash)
}

/// Deriva una chiave di crittografia dal codice della chat usando HKDF
#[derive(Zeroize, ZeroizeOnDrop)]
pub struct ChatKey {
    #[zeroize(skip)]
    cipher: ChaCha20Poly1305,
}

impl ChatKey {
    /// Deriva la chiave dal codice della chat (supporta sia formato numerico che base64)
    pub fn derive_from_code(chat_code: &str) -> Result<Self, CryptoError> {
        let chat_secret = if chat_code.len() == 6 && chat_code.chars().all(|c| c.is_ascii_digit()) {
            // Formato numerico: espandi a 32 byte usando HKDF
            let numeric_bytes = chat_code.as_bytes();
            let hkdf = Hkdf::<Sha256>::new(Some(b"rchat-numeric-code"), numeric_bytes);
            let mut expanded = [0u8; 32];
            hkdf.expand(b"rchat-secret-expansion", &mut expanded)
                .map_err(|_| CryptoError::InvalidChatCode)?;
            expanded.to_vec()
        } else {
            // Formato base64: decodifica direttamente
            let decoded = URL_SAFE_NO_PAD
                .decode(chat_code)
                .map_err(|_| CryptoError::InvalidChatCode)?;
            
            if decoded.len() != 32 {
                return Err(CryptoError::InvalidChatCode);
            }
            decoded
        };

        // Usa HKDF per derivare una chiave di crittografia
        let hkdf = Hkdf::<Sha256>::new(None, &chat_secret);
        let mut key_bytes = [0u8; 32];
        hkdf.expand(b"rchat-e2ee-v1", &mut key_bytes)
            .map_err(|_| CryptoError::KeyDerivationFailed)?;

        let cipher = ChaCha20Poly1305::new_from_slice(&key_bytes)
            .map_err(|_| CryptoError::KeyDerivationFailed)?;

        // Zeroizza i byte della chiave
        key_bytes.zeroize();

        Ok(Self { cipher })
    }

    /// Cripta un payload
    pub fn encrypt(&self, plaintext: &[u8]) -> Result<Vec<u8>, CryptoError> {
        // Genera un nonce random (96-bit per ChaCha20Poly1305)
        let mut nonce_bytes = [0u8; 12];
        OsRng.fill_bytes(&mut nonce_bytes);
        let nonce = Nonce::from(nonce_bytes);

        // Cripta
        let ciphertext = self
            .cipher
            .encrypt(&nonce, plaintext)
            .map_err(|_| CryptoError::EncryptionFailed)?;

        // Concatena nonce + ciphertext
        let mut result = nonce_bytes.to_vec();
        result.extend_from_slice(&ciphertext);

        Ok(result)
    }

    /// Decripta un payload
    pub fn decrypt(&self, encrypted: &[u8]) -> Result<Vec<u8>, CryptoError> {
        if encrypted.len() < 12 {
            return Err(CryptoError::DecryptionFailed);
        }

        // Estrai nonce e ciphertext
        let (nonce_bytes, ciphertext) = encrypted.split_at(12);
        let nonce = Nonce::from_slice(nonce_bytes);

        // Decripta
        self.cipher
            .decrypt(nonce, ciphertext)
            .map_err(|_| CryptoError::DecryptionFailed)
    }
}

#[derive(Debug, Clone)]
pub enum CryptoError {
    InvalidChatCode,
    KeyDerivationFailed,
    EncryptionFailed,
    DecryptionFailed,
}

impl std::fmt::Display for CryptoError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            CryptoError::InvalidChatCode => write!(f, "Codice chat non valido"),
            CryptoError::KeyDerivationFailed => write!(f, "Fallita derivazione chiave"),
            CryptoError::EncryptionFailed => write!(f, "Crittografia fallita"),
            CryptoError::DecryptionFailed => write!(f, "Decrittografia fallita"),
        }
    }
}

impl std::error::Error for CryptoError {}

File: python-qt-client/rchat/crypto.py
"""
Modulo crittografico per RChat - Implementazione Python pura
Compatibile al 100% con l'implementazione Rust del client terminal
Usa Argon2id per key derivation e XChaCha20Poly1305 per encryption (identico a Rust)
"""

import os
import hashlib
import base64
import secrets
from typing import Optional

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey
from cryptography.exceptions import InvalidSignature
from argon2.low_level import hash_secret_raw, Type
import nacl.secret
import nacl.utils
import blake3


def generate_chat_code() -> str:
    """Genera codice chat 512-bit base64url (compatibile con Rust)"""
    random_bytes = secrets.token_bytes(64)  # 512 bit
    return base64.urlsafe_b64encode(random_bytes).decode('ascii').rstrip('=')


def generate_numeric_chat_code() -> str:
    """Genera codice chat numerico a 6 cifre"""
    return f"{secrets.randbelow(900000) + 100000:06d}"


def chat_code_to_room_id(chat_code: str) -> str:
    """
    Genera room ID da chat code usando BLAKE3 + SHA3-512
    IDENTICO all'implementazione Rust
    """
    # Prima passata: BLAKE3 (IDENTICO a Rust)
    blake3_hasher = blake3.blake3()
    blake3_hasher.update(b"rchat-room-id-v2:")
    blake3_hasher.update(chat_code.encode('utf-8'))
    blake3_hash = blake3_hasher.digest()
    
    # Seconda passata: SHA3-512
    sha3_hash = hashlib.sha3_512(
        b"rchat-double-hash:" + blake3_hash
    ).digest()
    
    return base64.urlsafe_b64encode(sha3_hash).decode('ascii').rstrip('=')


class ChatKey:
    """
    Chiave di crittografia E2EE derivata dal chat code
    Usa XChaCha20Poly1305 (identico a Rust) con nonce 24 byte
    """
    
    def __init__(self, key: bytes):
        if len(key) != 32:
            raise ValueError("Key must be 32 bytes")
        # Usa XChaCha20Poly1305 compatibile con Rust (nonce 24 byte)
        self.cipher = nacl.secret.SecretBox(key)
    
    @classmethod
    def derive_from_code(cls, chat_code: str) -> 'ChatKey':
        """
        Deriva chiave dal chat code usando Argon2id
        IDENTICO all'implementazione Rust per compatibilità completa
        """
        # Step 1: Decodifica/espande chat code (IDENTICO a Rust)
        if len(chat_code) == 6 and chat_code.isdigit():
            # Codice numerico: espandi a 64 byte usando Argon2id con parametri specifici
            numeric_bytes = chat_code.encode('utf-8')
            
            # Prima derivazione per espandere codice numerico (IDENTICO a Rust)
            # Rust: m_cost=65536, t_cost=3, p_cost=4, output_len=64
            salt = b"rchat-numeric-salt-v2-extreme"
            chat_secret = hash_secret_raw(
                secret=numeric_bytes,
                salt=salt,
                time_cost=3,           # t_cost = 3 iterazioni
                memory_cost=65536,     # m_cost = 64 MiB
                parallelism=4,         # p_cost = 4 thread
                hash_len=64,           # 512-bit output
                type=Type.ID,          # Argon2id
                version=19             # Argon2 v0x13
            )
        else:
            # Codice base64: decodifica e verifica 512-bit
            try:
                padding = '=' * (4 - len(chat_code) % 4) if len(chat_code) % 4 else ''
                chat_secret = base64.urlsafe_b64decode(chat_code + padding)
                if len(chat_secret) != 64:
                    raise ValueError("Invalid chat code length")
            except Exception:
                raise ValueError("Invalid chat code format")
        
        # Step 2: Deriva salt usando BLAKE3 del segreto (IDENTICO a Rust)
        salt_hasher = blake3.blake3()
        salt_hasher.update(b"rchat-e2ee-v2-salt:")
        salt_hasher.update(chat_secret)
        salt_hash = salt_hasher.digest()
        salt = salt_hash[:32]  # Primi 32 byte come salt
        
        # Step 3: Deriva chiave finale con parametri ad alta sicurezza (IDENTICO a Rust)
        # Rust: m_cost=131072, t_cost=4, p_cost=8, output_len=32
        key = hash_secret_raw(
            secret=chat_secret,
            salt=salt,
            time_cost=4,           # t_cost = 4 iterazioni (NON 3!)
            memory_cost=131072,    # m_cost = 128 MiB (NON 64!)
            parallelism=8,         # p_cost = 8 thread (NON 4!)
            hash_len=32,           # 256 bit output per XChaCha20
            type=Type.ID,          # Argon2id
            version=19             # Argon2 v0x13
        )
        
        return cls(key)
    
    def encrypt(self, plaintext: bytes) -> bytes:
        """Cripta con XChaCha20-Poly1305 (nonce 24 byte, compatibile Rust)"""
        nonce = nacl.utils.random(nacl.secret.SecretBox.NONCE_SIZE)  # 24 byte
        ciphertext = self.cipher.encrypt(plaintext, nonce)
        # PyNaCl encrypt() ritorna nonce+ciphertext+tag, ma vogliamo solo ciphertext+tag
        # Estrai solo la parte ciphertext (senza il nonce prepended)
        return nonce + ciphertext.ciphertext
    
    def decrypt(self, encrypted: bytes) -> bytes:
        """Decripta e verifica autenticità"""
        if len(encrypted) < 24:
            raise ValueError("Invalid ciphertext")
        
        nonce = encrypted[:24]
        ciphertext = encrypted[24:]
        
        # PyNaCl decrypt() vuole nonce+ciphertext insieme
        return self.cipher.decrypt(ciphertext, nonce)
    
    def encrypt_with_chain(self, plaintext: bytes, chain_key: bytes) -> bytes:
        """
        Cripta usando chain key (forward secrecy)
        Usa XChaCha20Poly1305 con nonce da 24 byte
        """
        if len(chain_key) != 32:
            raise ValueError("Chain key must be 32 bytes")
        
        temp_cipher = nacl.secret.SecretBox(chain_key)
        nonce = nacl.utils.random(nacl.secret.SecretBox.NONCE_SIZE)  # 24 byte
        ciphertext = temp_cipher.encrypt(plaintext, nonce)
        return nonce + ciphertext.ciphertext
    
    def decrypt_with_chain(self, encrypted: bytes, chain_key: bytes) -> bytes:
        """Decripta usando chain key"""
        if len(chain_key) != 32:
            raise ValueError("Chain key must be 32 bytes")
        if len(encrypted) < 24:
            raise ValueError("Invalid ciphertext")
        
        nonce = encrypted[:24]
        ciphertext = encrypted[24:]
        temp_cipher = nacl.secret.SecretBox(chain_key)
        return temp_cipher.decrypt(ciphertext, nonce)



class IdentityKey:
    """Chiave di identità Ed25519 per firme"""
    
    def __init__(self, private_key: Ed25519PrivateKey):
        self.private_key = private_key
        self.public_key = private_key.public_key()
    
    @classmethod
    def generate(cls) -> 'IdentityKey':
        """Genera nuova coppia di chiavi Ed25519"""
        private_key = Ed25519PrivateKey.generate()
        return cls(private_key)
    
    @classmethod
    def from_public_bytes(cls, public_key_bytes: bytes) -> 'IdentityKey':
        """Crea IdentityKey solo con chiave pubblica (per verifica)"""
        if len(public_key_bytes) != 32:
            raise ValueError("Public key must be 32 bytes")
        
        instance = cls.__new__(cls)
        instance.private_key = None
        instance.public_key = Ed25519PublicKey.from_public_bytes(public_key_bytes)
        return instance
    
    def public_key_bytes(self) -> bytes:
        """Ottieni la chiave pubblica come bytes (32 bytes)"""
        from cryptography.hazmat.primitives import serialization
        return self.public_key.public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw
        )
    
    def sign(self, message: bytes) -> bytes:
        """Firma un messaggio (64 bytes)"""
        if self.private_key is None:
            raise ValueError("Cannot sign without private key")
        return self.private_key.sign(message)
    
    def verify(self, message: bytes, signature: bytes) -> bool:
        """Verifica una firma - ritorna True se valida, False altrimenti"""
        if len(signature) != 64:
            return False
        
        try:
            self.public_key.verify(signature, message)
            return True
        except Exception:
            return False


class ChainKey:
    """Chain key per forward secrecy"""
    
    def __init__(self, initial_key: bytes):
        if len(initial_key) != 32:
            raise ValueError("Initial key must be 32 bytes")
        self.key = bytearray(initial_key)
        self.index = 0
    
    @classmethod
    def from_chat_code(cls, chat_code: str) -> 'ChainKey':
        """
        Inizializza chain dal codice chat usando Argon2id
        IDENTICO all'implementazione Rust (usa derive_key_material)
        """
        chat_code_bytes = chat_code.encode('utf-8')
        salt = b"chain-key-init"  # IDENTICO al salt Rust
        
        # Usa Argon2id con parametri IDENTICI a derive_key_material Rust
        # Rust: m_cost=128*1024, t_cost=4, p_cost=8
        base_key = hash_secret_raw(
            secret=chat_code_bytes,
            salt=salt,
            time_cost=4,           # t_cost = 4 (IDENTICO a Rust)
            memory_cost=131072,    # m_cost = 128 MiB (128*1024)
            parallelism=8,         # p_cost = 8 thread
            hash_len=32,
            type=Type.ID,
            version=19
        )
        
        return cls(base_key)
    
    def next(self) -> bytes:
        """Deriva prossima chiave nella chain (forward secrecy) - IDENTICO a Rust"""
        # Usa BLAKE3 come Rust per KDF ratcheting
        hasher = blake3.blake3()
        hasher.update(b"rchat-chain-ratchet:")
        hasher.update(bytes(self.key))
        hasher.update(self.index.to_bytes(8, 'little'))
        
        new_key_bytes = hasher.digest()
        new_key = new_key_bytes[:32]  # Primi 32 byte
        
        self.key = bytearray(new_key)
        self.index += 1
        
        return bytes(new_key)
    
    def advance_to(self, target_index: int):
        """Avanza a un indice specifico (ricrea da zero se target < current)"""
        if target_index < self.index:
            # Non possiamo tornare indietro, dobbiamo ricreare
            # Questo è un problema - dovremmo mantenere lo stato iniziale
            # Per ora, avanza solo in avanti
            return
        while self.index < target_index:
            self.next()
    
    def get_key_at_index(self, target_index: int) -> bytes:
        """
        Ottiene la chiave a un indice specifico senza modificare lo stato corrente
        Usa una chain temporanea
        """
        temp_chain = self.clone()
        if target_index < temp_chain.index:
            # Non possiamo tornare indietro con la chain esistente
            # Dobbiamo ricreare da zero - ma non abbiamo il chat_code qui!
            raise ValueError(f"Cannot get key at index {target_index} when current index is {temp_chain.index}")
        
        temp_chain.advance_to(target_index)
        return temp_chain.next()
    
    def clone(self) -> 'ChainKey':
        """Crea una copia della chain"""
        new_chain = ChainKey.__new__(ChainKey)
        new_chain.key = bytearray(self.key)
        new_chain.index = self.index
        return new_chain
